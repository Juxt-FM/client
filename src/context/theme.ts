/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import React from "react";

interface IThemeContext {
  colors: any;
  breakpoints: any;
}

export const ThemeContext = React.createContext<IThemeContext | undefined>(
  undefined
);

export function useTheme() {
  const theme = React.useContext(ThemeContext);
  return theme;
}

export function useCurrentBreakpoint() {
  const theme = useTheme();
  const [currentBreakpoint, setCurrentBreakpoint] = React.useState("lg");

  React.useEffect(() => {
    maybeChangeLayout();

    window.addEventListener("resize", maybeChangeLayout);
    return () => {
      window.removeEventListener("resize", maybeChangeLayout);
    };
  });

  const maybeChangeLayout = () => {
    const { breakpoints } = theme;
    if (window.innerWidth < breakpoints.sm) setCurrentBreakpoint("xs");
    else if (window.innerWidth < breakpoints.md) setCurrentBreakpoint("sm");
    else if (window.innerWidth < breakpoints.lg) setCurrentBreakpoint("md");
    else if (window.innerWidth < breakpoints.xl) setCurrentBreakpoint("lg");
    else setCurrentBreakpoint("xl");
  };

  return currentBreakpoint;
}
