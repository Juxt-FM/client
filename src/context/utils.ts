/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { useEffect, useRef } from "react";

/**
 * Returns the previous value of the
 * provided state
 * @param val
 */
export function usePreviousValue(val: any) {
  const ref = useRef(null);

  useEffect(() => {
    ref.current = val;
  }, [val]);

  return ref.current;
}
