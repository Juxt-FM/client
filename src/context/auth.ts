/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { useEffect } from "react";
import { useApolloClient, useLazyQuery } from "@apollo/client";
import { useSelector, useDispatch } from "react-redux";
import Cookies from "universal-cookie";
import jwt_decode from "jwt-decode";

import {
  FETCH_USER,
  FETCH_USER_FAIL,
  FETCH_USER_SUCCESS,
  LOGIN_USER,
  LOGOUT_USER,
  REFRESH_TOKEN,
  REFRESH_TOKEN_FAIL,
  REFRESH_TOKEN_SUCCESS,
  RootState,
} from "../store";

import {
  QUERY_AUTH_USER,
  MUTATION_REFRESH_TOKEN,
  User,
  setAccessToken,
  lockApollo,
  unlockApollo,
  createApolloClient,
  AuthCredentials,
} from "../graphql";
import { usePreviousValue } from "./utils";
import { useRouter } from "next/router";

const cookies = new Cookies();

const buildTokenInfo = (token: string) => {
  const decoded: any = jwt_decode(token);
  return {
    expires: decoded.exp,
    issued: decoded.iat,
  };
};

const refreshToken = () => {
  const client = createApolloClient({ standalone: true });

  // eslint-disable-next-line no-async-promise-executor
  return new Promise(async (resolve, reject) => {
    try {
      const { data } = await client.mutate({
        mutation: MUTATION_REFRESH_TOKEN,
      });
      if (data) {
        resolve(data.refreshToken);
      } else {
        reject(new Error());
      }
    } catch (e) {
      reject(e);
    }
  });
};

// Returns whether or not the user is logged in
export const useAuthStatus = () => cookies.get("logged_in") === "true";

interface IProvider {
  children: JSX.Element;
}

export const AuthProvider = ({ children }: IProvider): JSX.Element => {
  const loggedIn = useAuthStatus();
  const { onLogout } = useAuthActions();

  const { token } = useSelector((state: RootState) => state.auth);

  const prevToken = usePreviousValue(token);

  const dispatch = useDispatch();

  const onRefreshTokenSuccess = ({ accessToken }: AuthCredentials) => {
    dispatch({
      type: REFRESH_TOKEN_SUCCESS,
      payload: buildTokenInfo(accessToken),
    });

    setAccessToken(accessToken);
    unlockApollo();
  };

  const onRefreshTokenFailed = () => {
    dispatch({ type: REFRESH_TOKEN_FAIL });

    unlockApollo();

    onLogout(false);
  };

  const refresh = () => {
    lockApollo();
    dispatch({ type: REFRESH_TOKEN });
    refreshToken().then(onRefreshTokenSuccess).catch(onRefreshTokenFailed);
  };

  useEffect(() => {
    if (loggedIn) refresh();
  }, []);

  /**
   * Refresh the user's token every 14:45 minutes.
   * need to make sure the token info is actually new
   */
  useEffect(() => {
    if (token && token !== prevToken) {
      const timeout = token.expires * 1000 - token.issued * 1000 - 15000;

      // tslint:disable-next-line:no-console
      console.log(`Refreshing token in ${timeout / 1000 / 60} minutes`);

      setTimeout(refresh, timeout);
    }
  }, [token]);

  return children;
};

export function useAuthActions() {
  const dispatch = useDispatch();
  const client = useApolloClient();
  const router = useRouter();

  const onLogin = (accessToken: string) => {
    dispatch({ type: LOGIN_USER, payload: buildTokenInfo(accessToken) });

    cookies.set("logged_in", "true", { path: "/" });
    setAccessToken(accessToken);

    client.clearStore().then(() => router.push("/"));
  };

  const onLogout = (redirectToLogin = true) => {
    dispatch({ type: LOGOUT_USER });

    setAccessToken(undefined);
    cookies.set("logged_in", undefined, { path: "/", expires: new Date() });

    client.clearStore().then(() => {
      if (redirectToLogin) router.push("/auth/login");
    });
  };

  return {
    onLogin,
    onLogout,
  };
}

export const UserProvider = ({ children }: IProvider) => {
  const loggedIn = useAuthStatus();

  const dispatch = useDispatch();

  const [fetchUser, { data, called }] = useLazyQuery<{
    me: User;
  }>(QUERY_AUTH_USER, {
    onError: (err) => {
      dispatch({ type: FETCH_USER_FAIL, payload: err });
    },
  });

  useEffect(() => {
    if (loggedIn) fetchUser();
  }, [loggedIn]);

  useEffect(() => {
    if (called) dispatch({ type: FETCH_USER });
  }, [called]);

  useEffect(() => {
    if (data) {
      dispatch({ type: FETCH_USER_SUCCESS, payload: data.me });
    }
  }, [data]);

  return children;
};

export const useAuthUser = () =>
  useSelector((state: RootState) => state.auth.user);
