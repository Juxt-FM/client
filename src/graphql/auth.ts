/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { setContext } from "@apollo/client/link/context";

// A user's access token, expires every 15 minutes
let accessToken: string | undefined;

let locked = false;

/**
 * Updates the access token in memory
 * @param newToken
 */
export const setAccessToken = (newToken: string | undefined) => {
  if (newToken !== accessToken) {
    // tslint:disable-next-line:no-console
    console.log("updating token");

    accessToken = newToken;
  }
};

/**
 * Lock the apollo client to make
 * all requests wait until it is unlocked
 */
export const lockApollo = () => {
  // tslint:disable-next-line:no-console
  console.log("locking");

  locked = true;
};

/**
 * Unlock apollo client, will initiate
 * all waiting requests
 */
export const unlockApollo = () => {
  // tslint:disable-next-line:no-console
  console.log("unlocking");

  locked = false;
};

/**
 * This injects the user's access token into the request headers,
 * this happens for every request
 *
 * We check to see if the client is unlocked before resolving the new headers.
 * If locked, use setTimeout to try again in 100ms. Else, resolve the headers
 */
export const authLink = setContext((_: any, { headers }: any) => {
  const retries = 0;
  const promise = () =>
    new Promise<any>((resolve) => {
      const buildContext = () => {
        if (!locked) {
          // tslint:disable-next-line:no-console
          console.log("apollo unlocked, fetching");

          if (accessToken) {
            resolve({
              headers: {
                ...headers,
                Authorization: `Bearer ${accessToken}`,
              },
            });
          } else {
            resolve(headers);
          }

          return;
        } else if (retries > 15) {
          return headers;
        } else {
          // tslint:disable-next-line:no-console
          console.log("apollo locked, waiting");

          setTimeout(buildContext, 100);
        }
      };

      buildContext();
    });

  return promise().then((authenticatedHeaders) => authenticatedHeaders);
});
