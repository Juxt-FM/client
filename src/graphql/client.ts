/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { useMemo } from "react";
import {
  ApolloClient,
  concat,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";
import { onError } from "apollo-link-error";
import merge from "deepmerge";

import cache from "./cache";

import { authLink } from "./auth";

export const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

const httpLink = new HttpLink({
  uri: "http://localhost:4000/graphql",
  credentials: "include",
});

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    for (const err of graphQLErrors) {
      switch (err.extensions.code) {
        default:
          // tslint:disable-next-line:no-console
          console.log(err);
      }
    }
  }
});

let apolloClient: ApolloClient<NormalizedCacheObject>;

interface IClientConfig {
  standalone: boolean;
}

export const createApolloClient = ({ standalone }: IClientConfig) => {
  if (standalone)
    return new ApolloClient<NormalizedCacheObject>({
      link: httpLink,
      cache: new InMemoryCache(),
    });

  const ssrMode = typeof window === "undefined";

  // @ts-ignore
  let link = concat(errorLink, concat(authLink, httpLink));

  // @ts-ignore
  if (ssrMode) link = concat(errorLink, httpLink);

  return new ApolloClient({
    ssrMode,
    link,
    cache,
  });
};

export function initializeApollo(initialState: any = null) {
  const _apolloClient =
    apolloClient ?? createApolloClient({ standalone: false });

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps

    const data = merge(initialState, existingCache, {
      // We are overwrtiting arrays on merge, by default they concatenate
      arrayMerge: (dest, source, options) => source,
    });

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function addApolloState(
  client: ApolloClient<NormalizedCacheObject>,
  pageProps: any
) {
  if (pageProps?.props) {
    pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract();
  }

  return pageProps;
}

export function useApollo(pageProps: any) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const store = useMemo(() => initializeApollo(state), [state]);
  return store;
}
