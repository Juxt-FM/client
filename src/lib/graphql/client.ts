/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import fetch from "cross-fetch";
import { useMemo } from "react";
import {
  ApolloClient,
  concat,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";
import { onError } from "apollo-link-error";
import merge from "deepmerge";

import cache from "./cache";

import { authLink } from "./auth";

export const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

const httpLink = new HttpLink({
  uri: "http://localhost:4000/graphql",
  credentials: "include",
  fetch,
});

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    for (const err of graphQLErrors) {
      switch (err.extensions.code) {
        default:
          // eslint-disable-next-line no-console
          console.log(err);
      }
    }
  }

  if (networkError) {
    // eslint-disable-next-line no-console
    console.log(networkError);
  }
});

let apolloClient: ApolloClient<NormalizedCacheObject>;

interface IClientConfig {
  standalone: boolean;
}

/**
 * Create a new Apollo Client instance
 *
 * @param {{standalone: boolean}}
 * @returns {ApolloClient}
 */
export const createApolloClient = ({ standalone }: IClientConfig) => {
  if (standalone)
    return new ApolloClient<NormalizedCacheObject>({
      link: httpLink,
      cache: new InMemoryCache(),
    });

  const ssrMode = typeof window === "undefined";

  // @ts-ignore
  let link = concat(errorLink, concat(authLink, httpLink));

  // @ts-ignore
  if (ssrMode) link = concat(errorLink, httpLink);

  return new ApolloClient({
    ssrMode,
    link,
    cache,
  });
};

/**
 * Initializes the application's Apollo Client instance.
 *
 * @param initialState
 * @returns {ApolloClient}
 */
export function initializeApollo(initialState: any = null) {
  const _apolloClient =
    apolloClient ?? createApolloClient({ standalone: false });

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps

    const data = merge(initialState, existingCache, {
      // We are overwrtiting arrays on merge, by default they concatenate
      arrayMerge: (dest, source) => source,
    });

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

/**
 * Merges the existing client cache with the new instance.
 *
 * @param client
 * @param pageProps
 */
export function addApolloState(
  client: ApolloClient<NormalizedCacheObject>,
  pageProps: any
) {
  if (pageProps?.props) {
    pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract();
  }

  return pageProps;
}

/**
 * Hook used to retrieve the Apollo Client instance
 * and merges existing cache.
 *
 * @param pageProps
 */
export function useApollo(pageProps: any) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const store = useMemo(() => initializeApollo(state), [state]);
  return store;
}
