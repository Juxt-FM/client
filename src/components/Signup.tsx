/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { useReducer, ChangeEvent } from "react";
import { useMutation } from "@apollo/client";
import Link from "next/link";

import { AuthCredentials, MUTATION_SIGNUP } from "../lib/apollo";
import { useAuthActions } from "../lib/context";

import { FormInput } from "./common/Inputs";

import { Button } from "./common/Buttons";
import { ContentSection, Header } from "./Authentication";

import styles from "../styles/modules/signup.module.scss";

const NAME_CHANGED = "name_changed";
const PASSWORD_CHANGED = "password_changed";
const CONFIRM_PASSWORD_CHANGED = "confirm_password_changed";
const EMAIL_CHANGED = "email_changed";
const PHONE_CHANGED = "phone_changed";

interface IAction {
  type: string;
  payload: string;
}

const initialFormState = {
  name: {
    value: "",
    required: false,
  },
  email: {
    value: "",
    required: true,
  },
  phoneNumber: {
    value: "",
    required: false,
  },
  password: {
    value: "",
    required: true,
  },
  confirmPassword: {
    value: "",
    required: true,
  },
};

function formReducer(state: typeof initialFormState, action: IAction) {
  switch (action.type) {
    case NAME_CHANGED:
      return { ...state, name: { ...state.name, value: action.payload } };
    case EMAIL_CHANGED:
      return { ...state, email: { ...state.email, value: action.payload } };
    case PHONE_CHANGED:
      return {
        ...state,
        phoneNumber: { ...state.phoneNumber, value: action.payload },
      };
    case PASSWORD_CHANGED:
      return {
        ...state,
        password: { ...state.password, value: action.payload },
      };
    case CONFIRM_PASSWORD_CHANGED:
      return {
        ...state,
        confirmPassword: { ...state.confirmPassword, value: action.payload },
      };
    default:
      return state;
  }
}

export const SignupForm = () => {
  const { onLogin } = useAuthActions();

  const [formValues, dispatch] = useReducer(formReducer, initialFormState);

  const getFormData = () => {
    const result = {};

    Object.keys(formValues).forEach((prop) => {
      // @ts-ignore
      const options = formValues[prop];

      if (options.value === "") {
        if (options.required)
          // @ts-ignore
          result[prop] = options.value;
      }
      // @ts-ignore
      else result[prop] = options.value;
    });

    return result;
  };

  const [signup, { loading, error }] = useMutation<{
    createUser: AuthCredentials;
  }>(MUTATION_SIGNUP, {
    variables: {
      data: getFormData(),
    },
    onCompleted: ({ createUser: { accessToken } }) => {
      onLogin(accessToken);
    },
    // tslint:disable-next-line:no-empty
    onError: () => {},
  });

  const onNameChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch({ type: NAME_CHANGED, payload: event.target.value });
  };

  const onEmailChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch({ type: EMAIL_CHANGED, payload: event.target.value });
  };

  const onPhoneChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch({ type: PHONE_CHANGED, payload: event.target.value });
  };

  const onPasswordChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch({ type: PASSWORD_CHANGED, payload: event.target.value });
  };

  const onConfirmPasswordChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch({ type: CONFIRM_PASSWORD_CHANGED, payload: event.target.value });
  };

  const getErrorsForField = (field: string): string | undefined => {
    if (!(error && error.graphQLErrors)) return undefined;

    const fieldErrors: string[] = [];

    error.graphQLErrors.forEach((err) => {
      const code = err.extensions.code;

      if (
        code === "BAD_USER_INPUT" &&
        err.extensions.invalidArgs.includes(field)
      )
        fieldErrors.push(err.message);
    });

    return fieldErrors[0];
  };

  const onSubmit = (e: any) => {
    e.preventDefault();
    signup();
  };

  return (
    <form onSubmit={onSubmit}>
      <div className={styles.signupForm}>
        <div className={styles.signupColumn}>
          <FormInput
            inputSize="sm"
            id="name"
            name="name"
            type="text"
            label="Full name"
            onChange={onNameChange}
            error={getErrorsForField("name")}
          />
          <FormInput
            inputSize="sm"
            id="email"
            name="email"
            type="email"
            label="Email address"
            onChange={onEmailChange}
            error={getErrorsForField("email.address")}
            required
          />
          <FormInput
            inputSize="sm"
            id="phone"
            name="phone"
            type="tel"
            label="Phone number"
            onChange={onPhoneChange}
            error={getErrorsForField("phone.number")}
          />
        </div>
        <div className={styles.signupColumn}>
          <p>Passwords must be longer than 8 characters.</p>
          <FormInput
            inputSize="sm"
            id="password"
            name="password"
            type="password"
            label="Password"
            onChange={onPasswordChange}
            error={getErrorsForField("password")}
            required
          />
          <FormInput
            inputSize="sm"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
            label="Confirm Password"
            onChange={onConfirmPasswordChange}
            error={getErrorsForField("confirmPassword")}
            required
          />
        </div>
      </div>

      <Button
        color="accent"
        label="Sign up"
        loading={loading}
        loadingLabel="Just a moment..."
        type="submit"
      />
    </form>
  );
};

export default function Signup() {
  return (
    <div className={styles.root}>
      <div className={styles.leftSignup}>
        <div className={styles.signupRoot}>
          <Header
            title="Create your account"
            description="Gain access to unlimited blog posts and real-time market analysis."
            color="accent"
          />
          <SignupForm />
          <div className={styles.options}>
            <div className={styles.switchAction}>
              <p>Already have an account?</p>
              <Link href="login">
                <a>Log in</a>
              </Link>
            </div>
          </div>
        </div>
      </div>
      <div className={styles.rightSignup}>
        <ContentSection
          title="Title 1"
          content="
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores
          praesentium similique aliquam culpa suscipit, itaque, id architecto
          nihil voluptatem consequatur et atque aspernatur ex totam cupiditate
          obcaecati quos autem sed."
        />
        <ContentSection
          title="Title 2"
          content="
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores
          praesentium similique aliquam culpa suscipit, itaque, id architecto
          nihil voluptatem consequatur et atque aspernatur ex totam cupiditate
          obcaecati quos autem sed."
        />
        <ContentSection
          title="Title 3"
          content="
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores
          praesentium similique aliquam culpa suscipit, itaque, id architecto
          nihil voluptatem consequatur et atque aspernatur ex totam cupiditate
          obcaecati quos autem sed."
        />
      </div>
    </div>
  );
}
