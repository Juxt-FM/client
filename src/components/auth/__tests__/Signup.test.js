/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { create, act } from "react-test-renderer";
import { MockedProvider } from "@apollo/client/testing";
import { GraphQLError } from "graphql";

import { useAuthActions } from "../../../lib/context";
import { MUTATION_SIGNUP } from "../../../lib/graphql";

import Signup from "../Signup";

jest.mock("../../../lib/context");

jest.mock("next/link", () => ({ children }) => children);

const mockToken = "access_token";

const mockInput = {
  email: "user@email.com",
  name: "Test User",
  password: "ab12cd34",
  confirmPassword: "ab12cd34",
};

const mocks = [
  {
    request: {
      query: MUTATION_SIGNUP,
      variables: { data: mockInput },
    },
    result: { data: { createUser: { accessToken: mockToken } } },
  },
];

const mockFieldErrors = [
  {
    request: {
      query: MUTATION_SIGNUP,
      variables: { data: mockInput },
    },
    result: { data: { createUser: { accessToken: mockToken } } },
  },
];

const setup = (mockData) => {
  const component = create(
    <MockedProvider mocks={mockData}>
      <Signup />
    </MockedProvider>
  );

  const form = component.root.findByProps({ id: "signupForm" });
  const email = component.root.findByProps({ id: "email" });
  const name = component.root.findByProps({ id: "name" });
  const password = component.root.findByProps({ id: "password" });
  const confirmPassword = component.root.findByProps({
    id: "confirmPassword",
  });
  const button = component.root.findByProps({ id: "signup" });

  return {
    form,
    email,
    name,
    confirmPassword,
    password,
    button,
    component,
  };
};

describe("<Signup />", () => {
  const onLogin = jest.fn();

  useAuthActions.mockReturnValue({ onLogin });

  it("renders without crashing", () => {
    const { component } = setup(mocks);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should signup user", async () => {
    const { form, email, name, password, confirmPassword, button } = setup(
      mocks
    );

    act(() => {
      email.props.onChange({ target: { value: mockInput.email } });
      name.props.onChange({ target: { value: mockInput.name } });
      password.props.onChange({ target: { value: mockInput.password } });
      confirmPassword.props.onChange({
        target: { value: mockInput.confirmPassword },
      });
    });

    act(() => {
      form.props.onSubmit({ preventDefault: jest.fn() });
    });

    expect(button.props.loading).toEqual(true);

    await act(async () => new Promise((resolve) => setTimeout(resolve, 0)));

    expect(button.props.loading).toEqual(false);
    expect(onLogin).toBeCalledWith(mockToken);
  });

  it("should throw field error", async () => {
    const { form, email, name, password, confirmPassword, button } = setup(
      mocks
    );

    act(() => {
      email.props.onChange({ target: { value: mockInput.email } });
      name.props.onChange({ target: { value: mockInput.name } });
      password.props.onChange({ target: { value: mockInput.password } });
      confirmPassword.props.onChange({
        target: { value: mockInput.confirmPassword },
      });
    });

    act(() => {
      form.props.onSubmit({ preventDefault: jest.fn() });
    });

    expect(button.props.loading).toEqual(true);

    await act(async () => new Promise((resolve) => setTimeout(resolve, 0)));

    expect(button.props.loading).toEqual(false);
    expect(onLogin).toBeCalledWith(mockToken);
  });
});
