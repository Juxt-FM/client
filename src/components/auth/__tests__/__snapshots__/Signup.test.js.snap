// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders without crashing 1`] = `
<div>
  <div>
    <div>
      <div>
        <a>
          <a>
            <img
              alt="brand logo"
              src="placeholder-file"
            />
          </a>
        </a>
        <h2>
          Create your account
        </h2>
        <p>
          Gain access to unlimited blog posts and real-time market analysis.
        </p>
      </div>
      <form
        id="signupForm"
        onSubmit={[Function]}
      >
        <div>
          <div>
            <div>
              <label
                htmlFor="name"
              >
                Full name
              </label>
              <input
                className=" "
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="Full name"
                type="text"
              />
            </div>
            <div>
              <label
                htmlFor="email"
              >
                Email address
              </label>
              <input
                className=" "
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="Email address"
                required={true}
                type="email"
              />
            </div>
            <div>
              <label
                htmlFor="phone"
              >
                Phone number
              </label>
              <input
                className=" "
                id="phone"
                name="phone"
                onChange={[Function]}
                placeholder="Phone number"
                type="tel"
              />
            </div>
          </div>
          <div>
            <p>
              Passwords must be longer than 8 characters.
            </p>
            <div>
              <label
                htmlFor="password"
              >
                Password
              </label>
              <input
                className=" "
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
              />
            </div>
            <div>
              <label
                htmlFor="confirmPassword"
              >
                Confirm Password
              </label>
              <input
                className=" "
                id="confirmPassword"
                name="confirmPassword"
                onChange={[Function]}
                placeholder="Confirm Password"
                required={true}
                type="password"
              />
            </div>
          </div>
        </div>
        <button
          disabled={false}
          id="signup"
          type="submit"
        >
          Sign up
        </button>
      </form>
      <div>
        <div>
          <p>
            Already have an account?
          </p>
          <a>
            <a>
              Log in
            </a>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div>
    <section>
      <h2>
        Title 1
      </h2>
      <p>
         Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores praesentium similique aliquam culpa suscipit, itaque, id architecto nihil voluptatem consequatur et atque aspernatur ex totam cupiditate obcaecati quos autem sed.
      </p>
    </section>
    <section>
      <h2>
        Title 2
      </h2>
      <p>
         Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores praesentium similique aliquam culpa suscipit, itaque, id architecto nihil voluptatem consequatur et atque aspernatur ex totam cupiditate obcaecati quos autem sed.
      </p>
    </section>
    <section>
      <h2>
        Title 3
      </h2>
      <p>
         Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores praesentium similique aliquam culpa suscipit, itaque, id architecto nihil voluptatem consequatur et atque aspernatur ex totam cupiditate obcaecati quos autem sed.
      </p>
    </section>
  </div>
</div>
`;
